### PR TITLE
fix(TT-409): bubble up protocol errors to the client

### PR DESCRIPTION
**Pull request by raymondk**
* Bump the version of ic-agent to 0.38.2 so that it stops retrying 429s indefinitely and bubbles up 429s
* Make the http gateway forward 429s from the node
* Add a test with a mock container

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -42,8 +42,8 @@ hyper-util = "0.1"
 
 ic-cdk = "0.13"
 ic-cdk-macros = "0.13"
-ic-agent = "0.37"
-ic-utils = "0.37"
+ic-agent = "0.38.2"
```
Comment (nathanosdev) [inline Cargo.toml:45]: We should use `0.38` for both of these crates, that will lower the possibility of clashes between crate versions for consumers of libraries in this repo.

File: packages/ic-http-gateway/tests/protocol_error_tests.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        .arg("Dockerfile")
+        .arg("--force-rm")
+        .arg("--tag")
+        .arg(format!("{IMAGE_NAME}:{IMAGE_TAG}"))
+        .arg(".")
+        .output()?;
+
+    if !output.status.success() {
+        eprintln!("stderr: {}", String::from_utf8(output.stderr)?);
+        return Err("unable to build no_expose_port:latest".into());
```
Comment (nathanosdev) [inline packages/ic-http-gateway/tests/protocol_error_tests.rs:31]:
```suggestion
        return Err("unable to build mock busy replica image".into());
```

File: packages/ic-http-gateway/tests/protocol_error_tests.rs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn test_rate_limiting_error() -> Result<(), Box<dyn std::error::Error>> {
+    let res = build_gateway_image();
```
Comment (nathanosdev) [inline packages/ic-http-gateway/tests/protocol_error_tests.rs:39]: Why not just bubble this error up to be consistent with how errors are handled in the rest of the test?
```rust
build_gateway_image()?;
```

File: packages/ic-http-gateway/tests/protocol_error_tests.rs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    let healthcheck_url = format!("{}/healthcheck", backend_base_url);
+    let response = Client::new().get(&healthcheck_url).send().await?;
+    assert_eq!(
+        response.status().as_u16(),
+        200,
+        "Expected to receive 200 from /healthcheck but received {}",
+        response.status().as_u16()
+    );
+
+    // Make a gatway
```
Comment (nathanosdev) [inline packages/ic-http-gateway/tests/protocol_error_tests.rs:64]:
```suggestion
    // Make a gateway
```

File: Cargo.toml (hunk 1 above)
Comment (raymondk) [reply Cargo.toml:45]: I believe there is a 0.38.1 version that doesn't include the change we need. But I'll double check.

Comment (raymondk) [reply Cargo.toml:45]: This is 0.38.0 release: https://github.com/dfinity/agent-rs/blob/729398880545a1f49a4c6286d0dd0c94254b236f/ic-agent/src/agent/mod.rs#L1187

Maybe we can bump it at the next minor version bump.

**Review by nathanosdev — APPROVED** [review]

**Pull request merged**